src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala

```diff
@@ -528,6 +534,162 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
       Nil)
   }
 
+  def withConcurrentCompactionTestSetup(toggle: Boolean)(closure: String => Unit): Unit = {
+    // in order to slow down the merge operation and increase the conflict probability,
+    // we use a UDF with sleep as the merge condition
+    spark.udf.register("slowMergeCondition", (condition: Boolean) => {
+      Thread.sleep(10)
+      condition
+    })
+
+    withTempDir { dir =>
+      withSQLConf(DeltaSQLConf.CONCURRENT_COMPACTION_CONFLICT_DETECTION.key -> toggle.toString,
+                  // we enable merge repartition before write to slow down the merge operation,
+                  // thus increasing conflict probability
+                  DeltaSQLConf.MERGE_REPARTITION_BEFORE_WRITE.key -> "true") {
+        closure(dir.getAbsolutePath)
+      }
+    }
+  }
+
+  val concurrentCompactionTestPartitions = 3
+  val concurrentCompactionTestRangeSize = 10
+  val concurrentCompactionTestNumIterations = 10
+
+  def createConcurrentCompactionTestTable(location: String): Unit = {
+    spark.range(0)
+      .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+      .coalesce(1)
+      .write.partitionBy("par").format("delta").save(location)
+  }
+
+  def executeConcurrentCompactionTestInsertOnlyMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+      .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = DeltaTable.forPath(spark, location)
+    table.alias("t").merge(df.alias("s"), "slowMergeCondition(s.par=t.par AND s.id=t.id)")
+      .whenNotMatched().insertAll()
+      .execute()
+  }
+
+  def executeConcurrentCompactionTestUpsertMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+      .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = DeltaTable.forPath(spark, location)
+    table.alias("t").merge(df.alias("s"), "slowMergeCondition(s.par=t.par AND s.id=t.id)")
+      .whenNotMatched().insertAll()
+      .whenMatched().updateAll()
+      .execute()
+  }
+
+  def assertConcurrentCompactionTestResults(location: String): Unit = {
+    val expectedDF = spark
+      .range(concurrentCompactionTestNumIterations + concurrentCompactionTestRangeSize - 1)
+      .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+      .orderBy("id")
+    val deltaResultDF = spark.read.format("delta").load(location).orderBy("id")
```

Comment:
```
import io.delta.tables.DeltaTable
DeltaTable.forPath(location).toDF.orderBy("id")
```

We're in the Delta Lake land so let's use the API as much as possible. The often we use it in tests the more it's tested, no?

Please make sure it's used throughout the PR.

Reply:
Aside from table reads, are there any other places where I should refactor to use the Delta Lake API?

For now, I identified a total of 2 table reads and refactored them both.